File: typescript/src/ships/fitted-module.ts
/**
 * Immutable fitted-module snapshots returned by {@link ShipLoadout}.
 *
 * @packageDocumentation
 */

import type { AmmunitionCapacity } from './ammunition.js';
import type { OutfittingModule } from './modules.js';
import type { PreEngineeredVariant } from './pre-engineered.js';
import type { LoadoutModule, ModuleEngineering } from './slef.js';

/**
 * A point-in-time, deeply frozen view of the module fitted in one slot.
 *
 * The view is detached from its {@link ShipLoadout}: later edits do not change it, and
 * mutating it throws. Fetch a new view with {@link ShipLoadout.fittedModuleAt} after a
 * state-changing edit. All mutations live on `ShipLoadout`, keyed by {@link slot}; this
 * avoids the stale-handle lifecycle that a live proxy would otherwise need.
 *
 * @example
 * ```ts
 * import type { ShipLoadout } from '@elite-dangerous-almanac/core/ships/ship-loadout';
 *
 * declare const build: ShipLoadout;
 *
 * const before = build.fittedModuleAt('FrameShiftDrive')!;
 * build.applyBlueprint(before.slot, 'FSD_LongRange', { grade: 5 });
 * const after = build.fittedModuleAt(before.slot)!;
 * before.engineering; // unchanged
 * after.engineering;  // the applied blueprint
 * ```
 */
export interface FittedModule {
    /** Slot key in the build's own spelling. */
    readonly slot: string;
    /** Frontier module symbol, e.g. `"Int_Hyperdrive_Size6_Class5"`. */
    readonly symbol: string;
    /** Whether the module was powered on, or `undefined` when unspecified. */
    readonly on: boolean | undefined;
    /** Zero-based power-priority group, or `undefined` when unspecified. */
    readonly priority: number | undefined;
    /** Module health in `[0, 1]`, or `undefined` when unspecified. */
    readonly health: number | undefined;
    /** Captured purchase value in credits, or `undefined` when unspecified. */
    readonly value: number | undefined;
    /** Applied engineering state; otherwise `undefined`. */
    readonly engineering: ModuleEngineering | undefined;
    /** Detached, journal-shaped fitted record. */
    readonly raw: LoadoutModule;
    /**
     * Snapshotted fitted-article stats before its journal modifier block is folded, or
     * `null` when unresolved.
     *
     * A stock or ordinarily engineered module exposes its base catalogue record. A fixed
     * pre-engineered variant exposes its resolved article record here as well as through
     * {@link effectiveStats}, so stats omitted by a journal capture still describe the
     * article. Clearing or replacing that fixed engineering restores the stock record
     * before applying the next recipe.
     */
    readonly stats: OutfittingModule | null;
    /**
     * Post-engineering module stats, or `null` when unresolved. For weapons, journal
     * damage per second is resolved back to per-round damage and falloff is capped at
     * maximum range. Exact damage components follow the engineered total and disappear
     * when a damage conversion replaces them with a fractional distribution.
     * A module engineered through {@link ShipLoadout.applyBlueprint} also retains
     * recipe-only burst values that its journal-shaped modifier block does not serialize.
     * A festive variant fitted through {@link ShipLoadout.setPreEngineeredVariant} uses
     * its fixed modifier block.
     */
    readonly effectiveStats: OutfittingModule | null;
    /** Fully rearmed ammunition capacity, or `null` for modules without ammunition. */
    readonly ammunition: AmmunitionCapacity | null;
    /** Identified fixed pre-engineered variant, or `null` when not uniquely identified. */
    readonly preEngineeredVariant: PreEngineeredVariant | null;
}

File: dotnet/src/EliteDangerousAlmanac/Ships/FittedModule.cs
namespace EliteDangerousAlmanac.Ships;

/// <summary>A point-in-time view of the module fitted in one mount.</summary>
/// <remarks>
/// The view is detached from the build it came from, so a later edit does not change it.
/// Read a new one after an edit. Every edit lives on the build itself and is keyed by
/// <see cref="Slot"/>, which is what keeps this value free of a stale-handle lifecycle.
/// </remarks>
/// <param name="Slot">The mount key, in the build's own spelling.</param>
/// <param name="Symbol">The Frontier module symbol, such as <c>Int_Hyperdrive_Size6_Class5</c>.</param>
/// <param name="Raw">The detached, journal-shaped fitted record.</param>
public sealed record FittedModule(string Slot, string Symbol, LoadoutModule Raw)
{
    /// <summary>Whether the module is powered on, or <see langword="null"/> when unstated.</summary>
    public bool? On { get; init; }

    /// <summary>The power-priority group, or <see langword="null"/> when unstated.</summary>
    public int? Priority { get; init; }

    /// <summary>The module's health, 0 through 1, or <see langword="null"/> when unstated.</summary>
    public double? Health { get; init; }

    /// <summary>The captured purchase value in credits, or <see langword="null"/> when unstated.</summary>
    public double? Value { get; init; }

    /// <summary>The applied engineering, or <see langword="null"/> when the module is stock.</summary>
    public ModuleEngineering? Engineering { get; init; }

    /// <summary>
    /// The fitted article's stats before its journal modifier block is folded in, or
    /// <see langword="null"/> when the article did not resolve.
    /// </summary>
    /// <remarks>
    /// A stock or ordinarily engineered module shows its base catalogue record. A fixed
    /// pre-engineered variant shows its resolved article record here as well as through
    /// <see cref="EffectiveStats"/>, so a stat a journal capture left out still describes the
    /// article. Clearing or replacing that fixed engineering restores the stock record before
    /// the next recipe is applied.
    /// </remarks>
    public OutfittingModule? Stats { get; init; }

    /// <summary>
    /// The post-engineering stats, or <see langword="null"/> when the article did not resolve.
    /// </summary>
    /// <remarks>
    /// <para>
    /// On a weapon, journal damage per second is resolved back to per-round damage and the
    /// falloff is capped at the maximum range. Exact damage components follow the engineered
    /// total, and they disappear when a damage conversion replaces them with a fractional
    /// distribution.
    /// </para>
    /// <para>
    /// A thruster's or a shield generator's mass curve moves as a whole: a recipe names the
    /// optimal mass and the optimal multiplier, and the four endpoints follow them. A
    /// generator's maximum mass is the exception, because a lightened optimum leaves it where
    /// it was. So this record carries the curve the build's own metrics read.
    /// </para>
    /// </remarks>
    public OutfittingModule? EffectiveStats { get; init; }

    /// <summary>
    /// The fully rearmed ammunition capacity, or <see langword="null"/> on a module that takes
    /// no ammunition.
    /// </summary>
    public AmmunitionCapacity? Ammunition { get; init; }

    /// <summary>
    /// The fixed pre-engineered variant this article is, or <see langword="null"/> when no one
    /// variant answers to it.
    /// </summary>
    public PreEngineeredVariant? PreEngineeredVariant { get; init; }
}
